### DISCUSSION TITLE
Strange console logs showing when running Tauri dev : BuyNow driver Initialized

### DISCUSSION THREAD

**Discussion by developGrace** [tauri-apps/tauri#9831] [Q&A]
I am new to using Tauri and I have started to see these logs in my console:

![image](https://github.com/tauri-apps/tauri/assets/88786976/86fb3cc7-de11-419c-a183-35614b5bedc0)

How do I find what is causing this and get rid of it? I have searched for the code in my project and cannot find anything. Any help or guidance is much appreciated!

It links to: debugger:///VM11

source code 

```
!function() {
    'use strict';
    let t = !1;
    function e() {
        window.BuyNowRuntime.postMessageToHost('StartBuyNow')
    }
    function o() {
        window.setTimeout((()=>{
            let t = null;
            try {
                t = document.getElementsByTagName('social-lead-gen-in-article')[1].shadowRoot.children[0].getElementsByClassName('button-content')[0]
            } catch {
                return void o()
            }
            t && (t.textContent = 'Subscribe Now',
            console.log('MSN parser Found button!!!!!'),
            t.addEventListener('click', (t=>{
                e(),
                t.stopPropagation()
            }
            ), !1))
        }
        ), 200)
    }
    window.ExecuteBuyNowDriver = function() {
        console.log('MsnDriver starts'),
        o()
    }
    ,
    window.OnBuyNowClick = e,
    window.BuyNowRuntime = new class {
        constructor() {
            this.onCheckoutFinished = void 0,
            this.orderDetails = void 0
        }
        raiseMessageFromHost(t) {
            const e = t.shift();
            e && 'StartMerchantAuth' === e ? console.log(`Received checkout results: ${t[0]}`) : e && 'CheckoutFinished' === e && this.onCheckoutFinished && this.onCheckoutFinished(t[0])
        }
        postMessageToHost(t, e) {
            try {
                walletBuyNowNativeHandler && walletBuyNowNativeHandler.sendMessageToHost(t, e)
            } catch (t) {
                console.log(t)
            }
        }
        initialize(e) {
            return t || (console.log('BuyNow driver Initialized!!!!!!'),
            window.ExecuteBuyNowDriver(),
            t = !0),
            !0
        }
        StartBuyNowRuntime(t) {
            this.onCheckoutFinished = t
        }
    }
}();
```

**Comment by ArcticFqx** [comment]
I am having the same issue, not with Tauri, but with Neutralino, so I think there is something that is injecting itself into WebView2. I noticed it the first time yesterday as well when debugging my project.

The same strings are not in my code or my dependencies. I searched GitHub and found nothing, until today when you posted this.

Some findings when I tried figuring this out, it only seems to appear when the current page is on `localhost`. When I tried with `127.0.0.1` as the domain instead, it stopped appearing. Hopefully this helps when tracking down what's going on.

**Reply by developGrace** [reply]
I started this project last week, and like you said I only just started seeing it today 5/20/24.

**Reply by devanmauch** [reply]
I'm having the same issue and I've got no idea what it is. Just reinstalled Intellij on a new PC and ran a live server.

**Comment by lukedays** [comment]
I saw these messages today while debugging a .NET app on Microsoft Edge. Seems to be some injected (potentially malicious) code.
Maybe worth a post on Hackernews/reddit? 

Edge version: 125.0.2535.51

**Reply by developGrace** [reply]
It seems like is designed for a very specific webpage :  t = document.getElementsByTagName('social-lead-gen-in-article')[1].shadowRoot.children[0].getElementsByClassName('button-content')[0]

My first worry was also that is was a malicious injection, although I am no expert on that topic what so ever...

**Reply by lukedays** [reply]
Very curious indeed...

**Reply by developGrace** [reply]
I created a hackernews account and posted : https://news.ycombinator.com/item?id=40419161

Fingers crossed someone figures something out I guess 🤞

**Reply by lukedays** [reply]
Oh I saw now you posted! Did the same
https://news.ycombinator.com/item?id=40419281

**Reply by gilmorem560** [reply]
I'm not so sure this is isolated to WebView, I was debugging an Angular application, no .NET to speak of, and had the same happening. I'm running a few more experiments with Edge to see if I can get this to happen again. It's not happening on my macOS machine but my tests on Windows thus far have just been navigating to random sites. I'm building an Angular app to debug in Edge on Windows as I type this.

**Reply by gilmorem560** [reply]
Ouch, indeed is does crop back up in my Windows environment. Problem appears endemic to MS Edge debug sessions of diverse kinds right now. Advising my team to not touch Edge for debugging for the time being.

**Reply by RyanMatCook** [reply]
I notice it on my Mac m1, working on a react app in edge. Doesnt show in chrome

**Reply by DeepSeaCatz** [reply]
I have the same console logs with npm local host. Doesn't appear when I get out of localhost or when hosted on a server.

**Reply by kevinbburns** [reply]
Linux host, Webstorm, Reactjs/Vite front-end, .Net core backend, the same message in my console when using Edge for debugging.

**Comment by gilmorem560** [comment]
Throwing on the pile, saw this while debugging an Angular application. Opened a new tab and without a site even loaded in it, was seeing these console messages. I wish I had made note of the exact version Edge was on, because I immediately applied the latest upgrade to 125.0.2535.51 and I'm actually not seeing these messages anymore.

@lukedays with a fresh browser session and Edge reporting version 125.0.2535.51, do you still see these messages, either in the context you initially encountered them in or otherwise on a random page? For the record, I only got these on macOS. Tried on a Windows box that already had the latest Edge, no such messages. Tried again on my mac after Edge upgraded, likewise no such messages. Don't have anywhere else to test currently.

**Reply by gilmorem560** [reply]
Hmm, looking at the logs now the suggestion that the applied upgrade was in place must've meant it already went in, my logs show 5/17/24 as the date of application of the upgrade, my mistake I thought it upgraded by virtue of my opening the version information and seeing the latest version applied. Strange.

Anywho, I can confirm at least I'm not seeing this anymore, but that doesn't mean whatever it is has disappeared. In any case, I'm wary of Edge until such time as this is more publicized and addressed by M$.

**Reply by lukedays** [reply]
Yes, I'm seeing on 4 different Windows machines at that same version. It seems to be related to the debugger, in one of the machines running Windows 11, the messages only show up when starting a VS/VSCode debugger session. Always when pointing at localhost, other pages don't show it.

**Comment by gilmorem560** [comment]
Another update, this may be tied to some sort of M$ service worker and/or integrations they're trying to shove into Bing and MSN. I cleared service workers related to these sites and while I once again got strange things loading as debugger:///VM* objects, they now contain:

```
!function() {
    "use strict";
    var e = {};
    let t;
    e.d = function(t, o) {
        for (var i in o)
            e.o(o, i) && !e.o(t, i) && Object.defineProperty(t, i, {
                enumerable: !0,
                get: o[i]
            })
    }
    ,
    e.o = function(e, t) {
        return Object.prototype.hasOwnProperty.call(e, t)
    }
    ;
    let o, i;
    !function(e) {
        e.IsBuyNowReady = "IsBuyNowReady",
        e.BuyNowSubscribeNow = "BuyNowSubscribeNow",
        e.BuyNowQuickAuthFinished = "BuyNowQuickAuthFinished",
        e.BuyNowQuickAuthFailed = "BuyNowQuickAuthFailed",
        e.LoadLocalCards = "LoadLocalCards",
        e.BuyNowIsReady = "BuyNowIsReady",
        e.BuyNowStartQuickAuth = "BuyNowStartQuickAuth",
        e.BuyNowCheckoutFinished = "BuyNowCheckoutFinished",
        e.LoadLocalCardsFinished = "LoadLocalCardsFinished"
    }(o || (o = {})),
    function(e) {
        e.EntryPointReady = "EntryPointReady",
        e.StartBuyNow = "StartBuyNow",
        e.VisitPublisher = "VisitPublisher",
        e.MerchantAuthSucceeded = "MerchantAuthSucceeded",
        e.MerchantAuthFailed = "MerchantAuthFailed",
        e.LoadLocalCards = "LoadLocalCards"
    }(i || (i = {}));
    const a = "BuyNowRuntime";
    window[a] = new class {
        notifyHostPage(e, t) {
            window.dispatchEvent(new CustomEvent(e,{
                detail: t
            }))
        }
        listenOnHostPageEvent(e, t) {
            window.addEventListener(e, (e=>{
                t(e)
            }
            ))
        }
        handleQuickAuthFinished(e) {
            if (e.detail) {
                const t = JSON.parse(e.detail);
                "skipped" === t.type ? window[a].postMessageToHost(i.MerchantAuthFailed, []) : "display" !== t.type || t.displayed ? "dismissed" === t.type && "credential_returned" === t.reason && window[a].postMessageToHost(i.MerchantAuthSucceeded, []) : window[a].postMessageToHost(i.MerchantAuthFailed, [])
            }
        }
        handleQuickAuthFailed(e) {
            window[a].postMessageToHost(i.MerchantAuthFailed, [])
        }
        handleIsBuyNowReady() {
            window[a].notifyHostPage(o.BuyNowIsReady, "Ready"),
            window[a].postMessageToHost(i.EntryPointReady, [])
        }
        handleBuyNowSubscribeNow(e) {
            window[a].postMessageToHost(i.StartBuyNow, [e.detail]),
            t = JSON.parse(e.detail)
        }
        handleLoadLocalCards() {
            window[a].postMessageToHost(i.LoadLocalCards, [])
        }
        raiseMessageFromHost(e) {
            const i = e.shift()
              , s = JSON.parse(e[0]).event;
            if (i && "StartMerchantAuth" === i)
                this.listenOnHostPageEvent(o.BuyNowQuickAuthFinished, this.handleQuickAuthFinished),
                this.listenOnHostPageEvent(o.BuyNowQuickAuthFailed, this.handleQuickAuthFailed),
                this.notifyHostPage(o.BuyNowStartQuickAuth, s);
            else if (i && "VisitPublisher" === i) {
                if (t && t.plans) {
                    if (t.originalSubLink)
                        return void window.open(t.originalSubLink, "_blank").focus();
                    let e = 0;
                    for (let o = 0; o < t.length; o += 1)
                        if (t.plans[o].isPromotion) {
                            e = o;
                            break
                        }
                    if (t.plans[e].url) {
                        const o = new URL(t.plans[e].url);
                        window.open(o.origin, "_blank").focus()
                    }
                }
            } else
                i && "CheckoutFinished" === i ? window[a].notifyHostPage(o.BuyNowCheckoutFinished, s) : i && "LocalCardsFinished" == i && window[a].notifyHostPage(o.LoadLocalCardsFinished, s)
        }
        postMessageToHost(e, t) {
            try {
                walletBuyNowNativeHandler && walletBuyNowNativeHandler.sendMessageToHost(e, t)
            } catch (e) {}
        }
        initialize(e) {
            return window.BuyNowInitialized || (this.listenOnHostPageEvent(o.IsBuyNowReady, this.handleIsBuyNowReady),
            this.listenOnHostPageEvent(o.BuyNowSubscribeNow, this.handleBuyNowSubscribeNow),
            this.listenOnHostPageEvent(o.LoadLocalCards, this.handleLoadLocalCards),
            window.BuyNowInitialized = !0),
            !0
        }
    }
}();
```

Which to me looks like an "evolution" of the above script, perhaps pointing towards this being something that snuck out in an incomplete state in a service worker and has since been upgraded. Still worrisome that it just kinda pops up...but anyone seeing this, especially on Bing and MSN pages, try clearing the service workers related to those and then refresh everything, you may see the log messages go away.

Also regarding this specific repo this discussion happens to be on, needless to say, it's probably nothing in their camp anyway but I'm glad I wasn't alone in seeing this confusing console logging today. Better safe than sorry!

**Comment by GCRev** [comment]
In EdgeWebView/125.0.2535.51 I was able to find matches in `msedge.dll`. It really looks like the entire script is embedded in there.

Starting at byte offset 250568137 (according to ripgrep): 
```javascript
!function(){'use strict';let t=!1;function e(){window.BuyNowRuntime.postMessageToHost('StartBuyNow')}function o(){window.setTimeout((()=>{let t=null;try{t=document.getElementsByTagName('social-lead-gen-in-article')[1].shadowRoot.children[0].getElementsByClassName('button-content')[0]}catch{return void o()}t&&(t.textContent='Subscribe Now',console.log('MSN parser Found button!!!!!'),t.addEventListener('click',(t=>{e(),t.stopPropagation()}),!1))}),200)}window.ExecuteBuyNowDriver=function(){console.log('MsnDriver starts'),o()},window.OnBuyNowClick=e,window.BuyNowRuntime=new class{constructor(){this.onCheckoutFinished=void 0,this.orderDetails=void 0}raiseMessageFromHost(t){const e=t.shift();e&&'StartMerchantAuth'===e?console.log(`Received checkout results: ${t[0]}`):e&&'CheckoutFinished'===e&&this.onCheckoutFinished&&this.onCheckoutFinished(t[0])}postMessageToHost(t,e){try{walletBuyNowNativeHandler&&walletBuyNowNativeHandler.sendMessageToHost(t,e)}catch(t){console.log(t)}}initialize(e){return t||(console.log('BuyNow driver Initialized!!!!!!'),window.ExecuteBuyNowDriver(),t=!0),!0}StartBuyNowRuntime(t){this.onCheckoutFinished=t}}}();
```
